### PR TITLE
OAuth client: refresh before re-authorizing, and discover before refreshing

### LINKED ISSUE #1318: Critical Token Refresh Bugs - Prevents Proactive Refresh
### Initial Checks

- [x] I confirm that I'm using the latest version of MCP Python SDK
- [x] I confirm that I searched for my issue in https://github.com/modelcontextprotocol/python-sdk/issues before opening this issue

### Description

tldr: Circular logic in `mcp/client/auth.py` creates two critical issues that prevent proactive token refresh; this would cause frequent re-auth for all users but it creates a third, worse issue for Clients that use TokenStorage (especially noticeable in a multi-user setup)

## Background 
In `mcp/client/auth.py` there are only 3 places where `OAuthContext.token_expiry_time` is set:
1. During init it is set to None
2. During full re-auth after receiving a 401 (in call to `_handle_token_response`)
3. When a token is deemed invalid in `is_token_valid` (at start of `async_auth_flow`) 

## Issue 1 - Token validity check treats None as Valid
The ultimate problem is that when `token_expiry_time` is None, a token will pass validity checks, per the below. 

```python
def is_token_valid(self) -> bool:
    """Check if current token is valid."""
    return bool(
        self.current_tokens
        and self.current_tokens.access_token
        and (not self.token_expiry_time or time.time() <= self.token_expiry_time)
#                  ^^^^^^^^^^^^^^^^^^ None makes it perpetually valid so never needs pro-active refresh
    )
```
This is probably intended behavior to enable None to represent 'perpetual tokens'.

It sets up 2 fatal problems.

## Issue 2 - Expired tokens get converted to 'perpetual tokens'
In `update_token_expiry` an expired token with a `token.expires` time of `0` is 'Falsey' which sets `token_expiry_time` to None.
```python
def update_token_expiry(self, token: OAuthToken) -> None:
    """Update token expiry time."""
    if token.expires_in:     # <============================ 0 evaluates to False
        self.token_expiry_time = time.time() + token.expires_in
    else:
        self.token_expiry_time = None  # <============================ Makes it perpetually valid
```

Per **Issue 1**, when `token_expiry_time` is None the token is always valid (i.e. perpetual).

## Undesirable Result - Only full re-auth can refresh a 'perpetual token' that goes stale
Because the token is now perpetual, it never gets proactively refreshed so it always requires a 401 and full re-auth to refresh.

## Issue 3 - Clients with storage always start with `token_expiry_time` set to None
The OAuth init does not check token expiry:
```python
async def _initialize(self) -> None:
    """Load stored tokens and client info."""
    self.context.current_tokens = await self.context.storage.get_tokens()
    self.context.client_info = await self.context.storage.get_client_info()
    self._initialized = True
```
This is often fine for Clients that persist.

For newly or frequently instantiated Clients that use Token Storage, `token_expiry_time` is None by default.

Due to **Issue 1** that None value cause the token to become perpetual.

Due to **Issue 2** the perpetual token is never checked for refresh.

As a result, tokens from storage (which are probably stale) are never checked prior to their first call. So they always trigger a 401 and revert to full re-auth.

## Proposed Fixes
Token update should check if token has an expires_in attribute and for an explicit None. This preserves perpetual tokens while treating expired tokens appropriately.

```python
  def update_token_expiry(self, token: OAuthToken) -> None:
      """Update token expiry time."""
      if hasattr(token, 'expires_in') and token.expires_in is not None: # <================ Fix
          self.token_expiry_time = time.time() + token.expires_in
      else:
          self.token_expiry_time = None
```

Also, the validity check should only work when `self.token_expiry_time=None` explicitly instead of when 'Falsey':

```python
def is_token_valid(self) -> bool:
    """Check if current token is valid."""
    return bool(
        self.current_tokens
        and self.current_tokens.access_token
        and (self.token_expiry_time is None or time.time() <= self.token_expiry_time) # <============== Fix
               #^^^^^^^^^^^^^^^^^^^^^^^ None can remain signal for perpetual token
    )
```

Finally, OAuthContext._initialize should call update_token_expiry so stored tokens aren't treated as perpetual by default:

```python
async _initialize(self):
    """Initialize and properly set token expiry from stored tokens."""
    self.context.current_tokens = await self.context.storage.get_tokens()
    self.context.client_info = await self.context.storage.get_client_info()
    
    # Fix: Update token expiry if tokens loaded from storage
    if self.context.current_tokens:
        self.context.update_token_expiry(self.context.current_tokens)
    self._initialized = True
```

### Example Code

```Python

```

### Python & MCP Python SDK

```Text
I'm on 1.12.4 but the code is still the same for 1.13.1
```

### LINKED ISSUE #3240: OAuth token refresh hits the wrong endpoint when the auth server lives under a path
Ran into this with a hosted MCP server whose authorization server isn't at the origin root — token endpoint is `https://host/oauth2/api/v1/token`, not `https://host/token`.

If a client starts up with a cached-but-expired access token (+ refresh token) and hasn't done discovery yet, `async_auth_flow` refreshes at the very top — before any 401/metadata discovery. So `oauth_metadata is None` and `_refresh_token` uses the fallback `urljoin(get_authorization_base_url(server_url), "/token")`, i.e. just `{scheme}://{netloc}/token`. That 404s, `_handle_refresh_response` clears the tokens, and the flow drops to full interactive auth — which a headless/gateway client can't do. So the server silently disconnects every time the access token expires (mine are 5 min, so… constantly).

Same path-stripping fallback is in `_get_token_endpoint`, `_perform_authorization_code_grant` (`/authorize`) and DCR (`/register`) — refresh is just the one that bites silently.

Repro (roughly):
- MCP server whose AS metadata puts `token_endpoint` under a path, not `{origin}/token`
- log in normally so tokens get cached
- let the access token expire (or clear the expiry), reconnect with a fresh provider
- watch the refresh POST go to `https://host/token` → 404 → "Token refresh failed" → tokens cleared → it tries to open a browser

Fix looks like: discover metadata before the eager refresh (or stop dropping the issuer path in the fallback). Happy to PR — have a branch that pulls the PRM/ASM discovery out of the 401 branch and runs it before the refresh.

(used some AI help digging into this)

### LINKED ISSUE #3250: OAuthClientProvider._initialize() breaks transparent refresh: missing update_token_expiry + missing oauth_metadata load
### Summary

When a client process restarts (or any time `OAuthClientProvider` is reconstructed), the SDK fails to transparently refresh expired access_tokens even when a valid `refresh_token` is on disk and the IdP would happily exchange it. Users are forced through an interactive OAuth re-auth on every process restart — even when the refresh_token is still valid for up to 15 days per the IdP's policy.

This affects every MCP server that issues short-lived access_tokens (~15 min) with longer-lived refresh_tokens — Fold MCP, Notion, GitHub PAT-rotated OAuth, any Hydra-style server, etc. — i.e. **the entire modern OAuth ecosystem**. The symptom is indistinguishable from the server revoking the refresh_token.

### Bug 1: `_initialize()` doesn't compute `token_expiry_time`

`_initialize()` loads `current_tokens` from storage but never calls `context.update_token_expiry(token)`. So `context.token_expiry_time` stays `None`.

Then `is_token_valid()`:

```python
def is_token_valid(self) -> bool:
    return bool(
        self.current_tokens
        and self.current_tokens.access_token
        and (not self.token_expiry_time or time.time() <= self.token_expiry_time)
    )
```

When `token_expiry_time is None`, the second clause is `not None or …` = `True`, so the function **unconditionally returns True** regardless of whether the access_token is expired by 1 second or 1 hour.

The refresh-on-expiry guard in `async_auth_flow`:

```python
if not self.context.is_token_valid() and self.context.can_refresh_token():
    refresh_request = await self._refresh_token()
    …
```

…never fires. Expired access_tokens are sent on every request, the server returns 401, and the user lands in the full-re-auth branch (`async_auth_flow` lines 514+) which forces interactive login.

This same fix is **already applied on the write path**: `set_tokens()` (the function called after a successful refresh) **does** call `update_token_expiry()`, and there's even a comment in `set_tokens` referencing "Fix A … OAuthTokens.expiresAt persistence" that describes the pattern. The read path (`_initialize`) just doesn't do the same thing.

### Bug 2: `_refresh_token()` builds the wrong endpoint URL when `oauth_metadata` isn't loaded

`_refresh_token()` picks the token endpoint like this:

```python
if self.context.oauth_metadata and self.context.oauth_metadata.token_endpoint:
    token_url = str(self.context.oauth_metadata.token_endpoint)  # pragma: no cover
else:
    auth_base_url = self.context.get_authorization_base_url(self.context.server_url)
    token_url = urljoin(auth_base_url, "/token")
```

For a server like `https://mcp.fold.money/mcp`, the fallback path produces `https://mcp.fold.money/token` — **404**. The correct endpoint for Hydra-style servers is `https://mcp.fold.money/oauth/token`.

`oauth_metadata` is normally populated via server discovery during the 401-handling flow (after a 401). But the refresh-on-expiry path runs **before** any 401 — it proactively refreshes when the token is expired, with no 401 yet. So `oauth_metadata` is never populated, and refresh fails silently with 404. The user then sees the 401 → re-auth loop as if the refresh_token itself were invalid.

The fix is for `_initialize()` to also load `oauth_metadata` from storage (e.g., via `storage.load_oauth_metadata()`, which the reference `HermesTokenStorage` implementation already provides).

### Reproduction

Any MCP client using `OAuthClientProvider` against an IdP with ~15 min access_tokens and a Hydra-style token endpoint.

```python
import asyncio, httpx
from mcp.client.auth.oauth2 import OAuthClientProvider
from mcp.shared.auth import OAuthClientMetadata, OAuthToken, OAuthClientInformationFull

# Suppose these came from persistent storage (Hermes's HermesTokenStorage
# or any conforming storage impl):
client_info = OAuthClientInformationFull.model_validate(client_info_dict)
current_tokens = OAuthToken.model_validate({
    "access_token": "...", "token_type": "Bearer", "expires_in": 900,
    "refresh_token": "...", "scope": "mcp:read",
})

class _Storage:
    async def get_tokens(self): return current_tokens
    async def set_tokens(self, t): current_tokens = t  # in-memory for repro
    async def get_client_info(self): return client_info

provider = OAuthClientProvider(
    server_url="https://mcp.fold.money/mcp",
    client_metadata=OAuthClientMetadata(...),
    storage=_Storage(),
)
await provider._initialize()
print(provider.context.is_token_valid())  # → True, even if access_token is expired
print(provider.context.token_expiry_time) # → None (Bug 1)

# Now suppose we manually trigger the refresh (mimicking async_auth_flow):
import time
# Force expires_in = 0 in the loaded token (the way HermesTokenStorage does it):
# …

# Refresh URL points to /token, not /oauth/token (Bug 2):
req = await provider._refresh_token()
print(req.url)  # → "https://mcp.fold.money/token" (404), not "/oauth/token"
```

### Fix

In `src/mcp/client/auth/oauth2.py`, modify `OAuthClientProvider._initialize()`:

```python
async def _initialize(self) -> None:
    """Load stored tokens and client info."""
    import asyncio as _asyncio
    self.context.current_tokens = await self.context.storage.get_tokens()
    self.context.client_info = await self.context.storage.get_client_info()
    # Fix bug 1: compute absolute expiry from the loaded token's
    # `expires_in`, mirroring what set_tokens() does on the write path.
    if self.context.current_tokens is not None:
        self.context.update_token_expiry(self.context.current_tokens)
    # Fix bug 2: load oauth_metadata if the storage supports it, so
    # _refresh_token() can find the correct token_endpoint without
    # having to wait for server discovery.
    loader = getattr(self.context.storage, "load_oauth_metadata", None)
    if callable(loader):
        try:
            meta = loader()
            if _asyncio.iscoroutine(meta):
                meta = await meta
            if meta is not None:
                self.context.oauth_metadata = meta  # type: ignore[assignment]
        except Exception:
            pass
    self._initialized = True
```

The reference storage (`HermesTokenStorage` in some downstream clients like `hermes-agent`) already implements `load_oauth_metadata()` returning `OAuthMetadata.model_validate(<contents of {server}.meta.json>)`. For SDK-provided storage classes that don't yet implement this, the `getattr` guard makes the second fix a no-op — bug 2 only manifests for downstream storage classes that already populate `.meta.json`.

### Live verification

Patched locally against `mcp==1.28.1` on macOS (Hermes agent 0.20.0). 16-minute live repro against `https://mcp.fold.money`:

1. Login via OAuth → fresh token, mtime T0.
2. Wait 15 min past access_token expiry → token on disk is stale, mtime still T0 (SDK never touched file).
3. Make MCP call with forced-expired access_token + fresh refresh_token.
4. **Without the fix**: SDK sends expired token, gets 401, falls through to re-auth (browser prompt).
5. **With the fix**: SDK calls `https://mcp.fold.money/oauth/token` with `grant_type=refresh_token`, gets HTTP 200 with fresh rotated pair, writes back to disk. MCP call returns real data (verified: `get_total_balance` → ₹146,656.35 across 4 accounts).

### AI disclosure

Drafted with AI assistance (GPT-class model). The bug analysis, code path tracing, fix design, and live verification were all done by a human reviewer who understood every line. The fix itself is 12 lines, two of which mirror the existing `set_tokens` write-path logic.

### PR DESCRIPTION
**Pull request by maxisbey**
On a 401, `OAuthClientProvider` now tries the stored refresh token (after rediscovery) before falling back to interactive authorization, and the pre-request refresh only runs once authorization-server metadata is known, so it never guesses `{origin}/token`. A restarted or headless client with a persisted refresh token recovers from an expired access token without a browser.

Closes #3240, closes #3250, closes #1318.

## Motivation and Context

The provider had two ways to obtain a token that each lacked half of what they needed: the pre-request branch could refresh but never ran discovery, and the 401 branch ran discovery but never tried the refresh token. After a restart nothing restores an expiry, so the pre-request branch is skipped, the stale bearer draws a 401, and the flow goes to the browser with a usable refresh token in hand — headless clients just fail. If an application forced the pre-request path by reporting the loaded token as expired, the refresh went to a path derived from the server origin, which 404s for any authorization server under a path, and the refresh token was then discarded. #1318 reported this a year ago; #3240 and #3250 are the same defect from two angles. The 401 branch did refresh in 1.10/1.11; #1071 moved that block ahead of the request instead of duplicating it and nothing tested the restart path.

## What changed

- 401 branch, Step 5: `if can_refresh_token(): refresh` → on success retry the request; otherwise (no refresh token, or the server rejected it) run the full grant as before.
- Pre-request refresh: additionally gated on `oauth_metadata is not None`. In-process behaviour is unchanged (metadata is cached after the first flow); on a cold start the request goes out and the 401 branch discovers, then refreshes.
- A failed refresh (either site) resets `_initialized` inside `_handle_refresh_response`, so the next request re-reads storage; a long-lived headless provider whose one refresh hit a transient error retries it rather than staying tokenless for the life of the process. Classifying failures (keep the refresh token on 5xx, clear only on `invalid_grant`) is left for the follow-up below.
- A fresh dynamic registration clears any held tokens (they were issued to a previous client and can't be refreshed by the new one).
- A CIMD `client_id` is portable across authorization servers, but tokens issued under it are not: when the issuer that becomes known (from PRM, or from AS metadata on the legacy no-PRM path) differs from the record's SEP-2352 stamp — or the record carries no stamp, so its tokens have no confirmed origin — the record is kept and re-stamped and its tokens (and any cached metadata from the previous server) are dropped before anything is refreshed (the TypeScript SDK's `discardIfIssuerMismatch`). The existing bound-credential discard and this rule now live in one `_apply_issuer_binding` method called at both points the issuer becomes known, replacing the two inline copies.

`src/` is +55/−41. No public API change, `TokenStorage` untouched. This is the TypeScript SDK's ordering (`auth()` refreshes before `startAuthorization`) and matches C#/Rust.

## How Has This Been Tested?

Interaction tests against the SDK's own AS+RS in process, each logging in as a "first process" and then connecting a fresh provider over the same storage: refresh on 401 after restart (no handlers wired); the same against an AS mounted under `/oauth2/v1` (bare `/token` 404s) proving the refresh goes to the advertised endpoint; a rejected 401-path refresh falling back to authorization; a headless provider retrying the refresh on its next connection after a failed one; a fresh registration not presenting tokens left from a previous client; and a CIMD client keeping its id but dropping its tokens on an issuer change. Each was mutation-checked against the line it pins. Unit tests pin that an expired token is not refreshed ahead of the request while metadata is unknown, and the CIMD rebinding on both discovery paths (stamped elsewhere, and unstamped). Manifest gains `client-auth:refresh:on-401` and `client-auth:refresh:discovered-endpoint`; the harness gains a path-prefixed-AS shim and `expire_access_token`.

Also driven over real sockets (uvicorn serving the SDK AS with 3 s access tokens, then under a prefix; a separate application process with file-backed storage run repeatedly): headless runs after expiry show `401 → PRM → ASM → POST …/token grant_type=refresh_token → 200`; with `tokens.json` rewritten to `expires_in=-1` and an app that derives expiry on load (the #3240 setup) there is no guessed `/token` request and the refresh lands on `/oauth2/v1/token`.

## Breaking Changes

None. Behavioural notes: a 401 with a refresh token in storage now produces one token-endpoint request before any redirect (a `scope=` challenge wider than the grant is then met by the 403 step-up on the retry, as in the TypeScript SDK). A cold start with an access token the server no longer accepts costs one 401 round trip before the refresh, unless the application restores both the expiry and the AS metadata itself. Against 2025-03-26-era servers that publish no metadata at all, in-process expiry now also goes through that 401 (and the failing well-known probes) instead of a single pre-request refresh; everywhere else in-process behaviour is unchanged.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update

## Checklist
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context

An earlier revision of this branch also folded in the #3256 recovery (expired DCR secret / `invalid_client` → re-register once) and a cold-start pre-request discovery mode. Review surfaced enough edge cases in both (403 step-up path, pre-registered credentials, tokens outliving their registration, a coarse-clock test knob) that they're better handled separately, so this PR is cut back to the refresh fix and the rest follows:

- #3256 (lapsed `client_secret_expires_at`, `invalid_client` handling incl. the 403 step-up exchange) — separate PR.
- Stored-token contract: an SDK-stamped `issuer` (and client binding) plus `issued_at` on the persisted token, following the `OAuthClientInformationFull.issuer` precedent, and a way for `TokenStorage` to forget tokens. Known exposures until then, stated rather than fixed here: (a) for pre-registered / unstamped `client_info` the 401-path refresh presents the stored refresh token (and client secret, as the code exchange already did) to whichever issuer the resource server's PRM now names — needs a token issuer stamp to detect; (b) `clear_tokens()` is memory-only, so a process that re-registers but cannot complete authorization leaves the previous client's `tokens.json` beside the new `client_info`, and the next process presents that refresh token under the new client once (rejected as `invalid_grant`) before authorizing; (c) verbatim storages get no real expiry across restarts. Refresh-failure classification and the `{server-origin}` fallback base used when PRM named an AS whose metadata could not be fetched belong in the same pass.
- No conformance scenario issues a refresh token today (modelcontextprotocol/conformance#139 is the nearest), so this isn't observable there yet.
- Supersedes #2492, #2875, #3263. (#1784 and #3248 restore expiry in `_initialize`, which this PR deliberately doesn't; the stored-token follow-up covers that.)

<sub>[AI Disclaimer](https://gist.github.com/maxisbey/6123d132484e4c533eab519a2800693d)</sub>

### CODE REVIEW

File: src/mcp/client/auth/oauth2.py [hunk 1]
```diff
@@ -741,10 +746,18 @@ async def async_auth_flow(self, request: httpx2.Request) -> AsyncGenerator[httpx
...
+                            # Held tokens belong to a previous client and cannot be refreshed by this one.
+                            self.context.clear_tokens()
+
+                    # Step 5: Refresh with the stored refresh token first (RFC 6749 §6); run the full
+                    # authorization only when there is none or the server rejects it.
+                    refreshed = False
+                    if self.context.can_refresh_token():
+                        refresh_response = yield await self._refresh_token()
+                        refreshed = await self._handle_refresh_response(refresh_response)
+                    if not refreshed:
```
[the remark this answers is not in the document]
Comment (maxisbey) [reply src/mcp/client/auth/oauth2.py:759]: Fixed in ab403241: `_handle_refresh_response` now resets `_initialized` on any failure, so the 401-path refresh site behaves like the pre-request one and the next request re-reads storage instead of leaving a long-lived headless provider tokenless. Pinned by `test_a_headless_client_whose_refresh_failed_retries_it_from_storage_on_the_next_connection`. Only clearing on `invalid_grant` and keeping the token on 5xx is the error-classification change listed with the follow-up in the description.

<sub>[AI Disclaimer](https://gist.github.com/maxisbey/6123d132484e4c533eab519a2800693d)</sub>

File: src/mcp/client/auth/oauth2.py [hunk 2]
```diff
@@ -741,10 +746,18 @@ async def async_auth_flow(self, request: httpx2.Request) -> AsyncGenerator[httpx
...
-                    await self._handle_token_response(token_response)
+                            # Held tokens belong to a previous client and cannot be refreshed by this one.
+                            self.context.clear_tokens()
+
+                    # Step 5: Refresh with the stored refresh token first (RFC 6749 §6); run the full
+                    # authorization only when there is none or the server rejects it.
+                    refreshed = False
+                    if self.context.can_refresh_token():
+                        refresh_response = yield await self._refresh_token()
+                        refreshed = await self._handle_refresh_response(refresh_response)
```
[the remark this answers is not in the document]
Comment (maxisbey) [reply src/mcp/client/auth/oauth2.py:758]: The CIMD half is handled in ab403241 + 76542f8c: when PRM names an issuer other than the one stamped on a CIMD record, the record is kept and re-stamped and its tokens (and the cached metadata) are dropped at the same point as the existing SEP-2352 guard, so no refresh token reaches the newly named server; `test_a_cimd_client_keeps_its_id_but_drops_its_tokens_when_the_authorization_server_changes` pins it. Unstamped / pre-registered records cannot be told apart from legitimately portable ones without an issuer stamp on the stored token itself, so that part is written up as a known exposure in the description and is what the stored-token follow-up is for.

On gating the Step-5 refresh on `oauth_metadata`: deliberately not done, because 2025-03-26-style servers that publish no metadata would then never refresh at all (the pre-request path is already gated). The manifest entry wording was narrowed instead, and using the PRM-named AS rather than the server origin as the fallback base is listed with the same follow-up.

<sub>[AI Disclaimer](https://gist.github.com/maxisbey/6123d132484e4c533eab519a2800693d)</sub>

File: src/mcp/client/auth/oauth2.py [hunk 3]
```diff
@@ -741,10 +745,33 @@ async def async_auth_flow(self, request: httpx2.Request) -> AsyncGenerator[httpx
...
+                        str(self.context.oauth_metadata.issuer) if self.context.oauth_metadata else None
+                    )
+                    if (
+                        client_info.client_id == self.context.client_metadata_url
+                        and current_issuer is not None
+                        and client_info.issuer not in (None, current_issuer)
+                    ):
+                        self.context.clear_tokens()
+                        client_info.issuer = current_issuer
+                        await self.context.storage.set_client_info(client_info)
```
[the remark this answers is not in the document]
Comment (maxisbey) [reply src/mcp/client/auth/oauth2.py]: Good catch. Moved in 76542f8c: the CIMD issuer-change handling now sits beside the bound-credentials guard right after PRM discovery and drops the cached metadata exactly as that guard does, so a failed rediscovery cannot leave the previous server endpoints in play under the new stamp. `test_cimd_record_is_restamped_and_its_tokens_and_cached_metadata_dropped_when_prm_names_a_new_issuer` pins the tokens, the metadata and the stamp.

<sub>[AI Disclaimer](https://gist.github.com/maxisbey/6123d132484e4c533eab519a2800693d)</sub>

**Comment by maxisbey** [comment]
Parking this for now. The underlying bug affects 1.x and 2.x equally and isn't security-driven, so it's been deprioritised behind current work; the branch stays around and the linked issues remain open. When it comes back, the plan is to settle how stored tokens record their issuer/client first, so the 401-path refresh can be gated on that instead of the interim checks discussed above.
